alternator: make number of backlog connections for server sockets configurable

ScyllaPiotr opened the pull request:
Add new config variable `alternator_listen_socket_backlog` that is subsequently passed as a parameter to calls to `listen()`. The parameter influences HTTP and HTTPS server sockets of Alternator. Before the patch, the default Seastar limit of up to 100 backlog connections was used.
The default value (for Alternator) is now 1024.

refs: scylladb/scylla-enterprise#5614
refs: https://github.com/scylladb/seastar/pull/3312

The fix is required by a customer, therefore it needs to be backported.

ScyllaPiotr committed:
alternator: make listen socket backlog configurable and live-updatable

Add new config variable `alternator_listen_socket_backlog` that controls
the backlog parameter passed to listen() for Alternator's HTTP and HTTPS
server sockets. The default value is 1024 (previously the Seastar default
of 100 was used).

The parameter is live-updatable: when changed at runtime, the new backlog
is applied to all active Alternator server sockets by re-invoking the
listen() syscall on the underlying file descriptors (supported by Linux).

This includes a Seastar change that adds set_listen_backlog() to
server_socket and http_server, allowing the listen backlog to be
updated on already-listening sockets.

Fixes #18021

ScyllaPiotr commented:
@xemul since you reviewed already a [very similar patch](https://github.com/scylladb/scylladb/pull/18810), please review this one too.

nyh commented on alternator/controller.cc:
```diff
@@ -137,7 +137,9 @@ future<> controller::start_server() {
                     &_memory_limiter.local().get_semaphore(),
-                    _config.max_concurrent_requests_per_shard);
+                    _config.max_concurrent_requests_per_shard,
+                    _config.alternator_listen_socket_backlog
```
Nitpick: I think the word "socket" in the new parameter's name is unnecessary and could be dropped -  I think "alternator_listen_backlog" is shorter and clear enough (for people familiar with the Unix TCP/IP API). It is **listen**()'s **backlog** argument.

nyh commented on alternator/server.cc:
```diff
@@ -593,22 +593,26 @@ server::server(executor& exec, service::storage_proxy& proxy, gms::gossiper& gos
 
+        listen_options socket_listen_opts {
+            .reuse_address = true, // to match behaviour when no listen_options param. is passed to _http_server.listen() and _https_server.listen()
+            .listen_backlog = listen_socket_backlog(),
```
You need to decide whether you want to make the effort to make the alternator listen backlog live-updateable or not. If not, then we shouldn't pretend it is live-updatable and it shouldn't be passed as updateable value at all. But if you do want to support live updates, you made two mistakes in doing that:
1. You are missing a `liveness::LiveUpdate` when defining this configuration parameter.
2. If this variable is ever changed, you'll need to call listen() again to change the backlog (this supposedly works on Linux (see https://stackoverflow.com/questions/64050281/can-backlog-value-that-is-passed-to-listen-call-be-modified-later-on-without) but may need a Seastar patch (?). Right now your code will simply ignore changes to this variable after the initial connection establishment. 

Ideally, you'd have a test that checks the new feature, and that test could also test live-update by temporarily setting the backlog to 1 or something (but you'll probably need to use injections for such a test to avoid accepting the new connections, so... ick - I won't shed too many tears if you test this only manually).

ScyllaPiotr replied:
I never meant for this to be live-updatable. I'll correct this by passing value.

mykaul replied:
Was it corrected? Also, we'd really like thing to be live-updatable, if possible.

ScyllaPiotr replied:
Yes, it's corrected.
I'll change into live-updatable then.

ScyllaPiotr commented:
@xemul please re-review

xemul reviewed (approved):
LGTM

denesb commented:
Conflicts, needs rebase.

nyh reviewed (approved):
Looks good. Please add fixes https://github.com/scylladb/scylladb/issues/18021 (and in general, always look for an issue to "Fixes" when sending a PR).

The default of 1024 makes sense to me. Apache Httpd and others which copied from it use 511, but there's no real reason not to go higher. The modern Linux maximum is 4096 (/proc/sys/net/core/somaxconn) so we're nowhere close.

Usually, I would ask you how you tested this code, but I have a vague memory that I already reviewed this patch a dozen times and was already sure it was correct :-)

ScyllaPiotr committed:
test: cluster: add test for alternator_listen_socket_backlog config

Verify that the alternator_listen_socket_backlog configuration option
correctly sets the OS socket backlog on both HTTP and HTTPS (TLS)
listeners, and that live-updating it via SIGHUP takes effect on both.
Uses ss(8) to read the Send-Q of the listen sockets, which on Linux
reflects the backlog passed to listen(2).

This also serves as a sanity check that reasonable backlog values
are successfully settable and accepted by the kernel without being
silently clamped or rejected.

ScyllaPiotr commented:
Moving to draft until the [seastar dependency](https://github.com/scylladb/seastar/pull/3312) makes it to its master.
